### DISCUSSION TITLE
単体テストの考え方/使い方 3.2〜3.3

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#95] [General]
- 3.2 単体テストのフレームワーク
- 3.3 テストケース間で共有するテストフィクスチャ

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by y-kaneda** [comment]
## 3.2 単体テストのフレームワーク
- テストフレームワーク、良し悪しがある
    - テストツールの選び方とかあまり真剣に比較したり考えたことないかも。デファクトスタンダード的なのをいつも選んでる。
- 「理想なのは、単体テストで表現していることが開発者だけでなく非開発者にも伝わることです。」良い

## 3.3 テスト・ケース間で共有するテスト・フィクスチャ(test fixture)
- 複数のテストケース間で複数のコードを共有する場合、そのコードがいつどうやって使われるのかを把握することが重要
  - RSpec の shared_context, let, let!, before とか気を付けないと混乱しがち
- 準備(Arrange)でテストデータを作るのが膨大なコード量となるので、そこをうまく共通化すべし
- (Class名と同じpublic関数がコンストラクタか)
#### Column テスト・フィクスチャ
- 決められた(fixed)状態に毎回なっているべきもの
  - 冪等でなければfixtureではない

### 3.3.1 コンストラクタの利用はテスト・ケース間の結び付きを強めてしまう
- 含まれるテストケース全てに同じ事前条件が強要されてしまう
  - このケースだけちょっとテストデータ変える、ができない

### 3.3.2 コンストラクタの利用はテスト・ケースの読みやすさを損わせてしまう
- 準備と実行のコードが離れるため、またコンストラクタは暗黙で呼ばれるため、意識していないと見誤る

### 3.3.3 テスト・フィクスチャに関するコードを共有するためのより良い方法
- プライベートなファクトリ・メソッドを導入する
  - 引数を取る、フィクスチャを作成する関数
- 事前のDB接続等、完全に共通なものはコンストラクタに書いて良い

**Comment by xxManatoxx** [comment]
## 3.2 単体テストのフレームワークについて

- 各テストケースはプロダクションコードが解決しようとしている物語（story）について語るべき
  - 語るべき物語
    - 問題領域に関する個別で不可分なシナリオ（もしくは事実）
  - テストが成功
    - シナリオ（または事実）が正しい証明
  - テストが失敗
    - 物語が事実ではない
      - 物語を書き直すか、プロダクションコードの修正を行う必要がある
  - 単体テストですべきこと
    - アプリケーションの振る舞いについてより高いレベルで描写すること
    - 理想
      - 開発者だけでなく、非開発者にも伝わること
  - 単体テストですべきでないこと
    - プロダクションコードが何をするのかをただ列挙するだけ

### 勉強になったポイント

- ここの解説ではBDDに通じる考え方を感じ、理想である非開発者にも伝わるテストコードを書けるようになりたいと感じた
  - これを目指すことで開発者同士で読みやすいテストコードを書くことに繋がると考えた

## 3.3 テストケース間で共有するテストフィクスチャ

- 複数のテスト間で同じコードを共有する場合
  - そのコードがいつどうやって使われるかを把握することが重要

### テストフィクスチャ

- テストフィクスチャは2つの意味で使われる
  - テストを実施する際に使われるオブジェクト
  - NUnit特有のものでTestFixture属性が付けられたクラスのこと
- コンストラクタ（またはsetup属性が付けられたメソッド）で共通処理を書く
  - 重大な欠点
    - テストケース間の結びつきが強くなってしまう
    - テストケースが読みづらくなってしまう

### 3.3.1 コンストラクタの利用はテストケース間の結びつきを強めてしまう

- コンストラクタでテストフィクスチャの準備を行う
  - 準備内で行っているフィクスチャの内容を変更すると一部のテストで失敗する可能性がでてくる
  - これは1つのテストケースの修正が他のテストケースに影響を与えてはならない、に背くことになる

### 3.3.2 コンストラクタの利用はテストケースの読みやすさを損なわせてしまう

- 準備のコードをコンストラクタに実装するもう一つの欠点
  - テストメソッドが読みにくくなる
    - 準備フェーズが他の部分に書かれているため、何を検証しようとしているかが理解しにくくなる
    - テストコード＋コンストラクタを見なくてはいけないから

### 3.3.3 テストフィクスチャに関するコードを共有するためのよりよい方法

- テストクラスにプライベートなファクトリメソッドを導入する
  - このファクトリメソッドが汎用的に作られていればテストケース間の結びつきは弱められる
  - ファクトリメソッドは中身を見なくても何をしているかをわかるようにする必要がある
    - リーダブルコードのお話
- テストフィクスチャのコンストラクタでの用意をしても良いケース
  - 全ての（もしくはほぼ全ての）テストケースで同じテストフィクスチャを使う場合

#### 勉強になったポイント

- 今まで既存のテストににならってSetUpメソッド（PHPUnit）を使用して初期データの準備をしていたが、そこを見直してみたいと感じた(とはいえBDDテスト導入して書き換えを検討しているため、実装し直す場合はガラッと変わりそう)
  - プライベートなファクトリメソッドとして定義できると思うのでどうしたら読みやすいテストになるか考えたい
  - SetUpメソッドで準備していた箇所は自分でも少し読みづらさを感じていたので実装したテストを振り返ってみる

**Comment by kentaro-yamaguchi** [comment]
## 3.2
1. 「テストケースはプロダクションコードが解決する物語を語るべき」というのは改めて念頭におきたい
1. 要件の理解が浅かったりすると虚偽の物語をどちらかで語ってしまうことがある。テスト失敗時にその虚偽の物語を真実と思い込みテストケースかプロダクションコードを改ざんしてしまうのがしばしば
## 3.3
1. 事前・事後準備はlet, before, after, shared_context等があるのでrspecはより宣言的でわかりやすく思う
1. テストコードに限らず共通化は不得意。後で「このケースとあのケース同じデータセットだった」と気づくことが多い
1. 「共通化にコンストラクタを使うべきでない」とは違うやもだがrspecでも感じる
    1. 「ここのケースでできるentityは1つなのに2つできてしまってる」という状況にぶつかり、rspecでいうfactoryやbeforeで共通化すべきでないものを生成してしまっていることがある
    1. 特にfactoryは全ケースで共有できるため、中で余計なものを作らないように気を付けたい

**Comment by myamamoto111** [comment]
## 3.2

- 単体テストのツール、フレームワークについて
  - 分かりやすくシンプルなものを選ぶ
- 単体テストを作成する際に考えること
  - プロダクションコードが何をするのかを単に列挙することではなく
  - アプリケーションの振る舞いについて高いレベルで描写すること
  - 理想は、開発者だけでなく、非開発者にも伝わること

## 3.3

- テスト・フィクスチャ
  - テストを実施する際に使われるオブジェクト
  - オブジェクトは各テストの前に決められた状態(fixed)状態に毎回なっていること
  - 同じテストを何回繰り返しても毎回同じ結果が得られること
- コンストラクタの利用
  - 1つのテストケースに関する修正が他のテストケースに影響を与えてはいけない
  - テストケースの読みやすさを損ねてしまわないように
  - コードを共有するためのより良い方法は、プライベートなファクトリメソッドで定義
  - 例外
    - 全てのテストケースで同じテストフィクスチャを使うような場合
      - DBの接続

**Comment by nmurakami0** [comment]
3.2

- 同一言語で確かにテストフレームワークは複数あるケースが多い
  - 人気の選べば大抵良さそう

3.3

- テストフィクスチャ準備の共通部分は基本的にコンストラクタではなくファクトリメソッドで作成する
  - rspecならshared_context、またはFactoryBotの複数モデルにまたがって変数を受け取るパターンが近そう
- 全テストで共通する部分はコンストラクタに書いて良い
  - Railsだと設定系のrbファイルに書くことも多そう (helper.rb的なものにとか)
